src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1589,30 +1614,34 @@ PlanRouterQuery(Query *originalQuery, RelationRestrictionContext *restrictionCon
 
 		shardsPresent = true;
 
-		/* all relations are now pruned down to 0 or 1 shards */
-		Assert(list_length(prunedShardList) <= 1);
-
-		shardInterval = (ShardInterval *) linitial(prunedShardList);
-
-		/* anchor shard id */
-		if (shardId == INVALID_SHARD_ID)
+		foreach(shardIntervalCell, prunedShardList)
 		{
-			shardId = shardInterval->shardId;
-		}
+			ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+			RelationShard *relationShard = CitusMakeNode(RelationShard);
+
+			relationShard->relationId = shardInterval->relationId;
+			relationShard->shardId = shardInterval->shardId;
 
-		/* add relation to shard mapping */
-		relationShard = CitusMakeNode(RelationShard);
-		relationShard->relationId = shardInterval->relationId;
-		relationShard->shardId = shardInterval->shardId;
+			/*
+			 * We need anchor shard id for compute tasks. Also, note that we have
+			 * single shard interval for select queries at this point. Therefore,
+			 * condition below should run once only for router select queries.
+			 */
+			if (!localMultiShardModifyQuery)
+			{
+				shardId = relationShard->shardId;
```

Comment:
I find it a bit hard to follow how the `if (!lockMultiShardModifyQuery)` relates to the comment. What if we extract the shardId separately? (e.g. introduce a `GetInitialShardId` function).